code-studio/src/js/components/GridEditor.jsx

```diff
@@ -0,0 +1,129 @@
+/**
+ * @overview React component to allow for easy editing of Karel Grids.
+ * Used in LevelBuilder, and relies on some apps code for validation.
+ * Supports both Bee and Farmer skins.
+ */
+/* global React, dashboard */
+
+var BeeCell = require('@cdo/apps/maze/beeCell');
+var Cell = require('@cdo/apps/maze/cell');
+var mazeUtils = require('@cdo/apps/maze/mazeUtils');
+
+var BeeCellEditor = require('./BeeCellEditor.jsx');
+var CellEditor = require('./CellEditor.jsx');
+var Grid = require('./Grid.jsx');
+
+window.dashboard = window.dashboard || {};
+
+var CellJSON = React.createClass({
+  propTypes: {
+    serialization: React.PropTypes.arrayOf(React.PropTypes.arrayOf(React.PropTypes.object)),
+    onChange: React.PropTypes.func.isRequired
+  },
+
+  componentDidUpdate: function () {
+    var node = this.refs.serializedInput.getDOMNode();
+    node.focus();
+    node.select();
+  },
+
+  onChange: function (event) {
+    this.props.onChange(JSON.parse(event.target.value));
+  },
+
+  render: function () {
+    return (<label>
+      Cell JSON (for copy/pasting):
+      <input type="text" value={JSON.stringify(this.props.serialization)} ref="serializedInput" onChange={this.onChange}/>
+    </label>);
+  }
+});
+
+window.dashboard.GridEditor = React.createClass({
+  propTypes: {
+    serializedMaze: React.PropTypes.arrayOf(React.PropTypes.arrayOf(React.PropTypes.object)),
+    maze: React.PropTypes.arrayOf(React.PropTypes.array), // maze items can be integers or strings
+    initialDirt: React.PropTypes.arrayOf(React.PropTypes.arrayOf(React.PropTypes.number)),
+    skin: React.PropTypes.string.isRequired,
+    onUpdate: React.PropTypes.func.isRequired
+  },
+
+  getInitialState: function () {
+    var cells;
+    var cellClass = this.getCellClass();
+
+    if (this.props.serializedMaze) {
+      cells = this.props.serializedMaze.map(function (row) {
+        return row.map(cellClass.deserialize);
+      });
+    } else {
+      cells = this.props.maze.map(function (row, x) {
+        return row.map(function (mazeCell, y) {
+          var initialDirtCell = this.props.initialDirt[x][y];
+          return cellClass.parseFromOldValues(mazeCell, initialDirtCell);
+        }, this);
+      }, this);
+    }
+
+    return {
+      cells: cells
+    };
+  },
+
+  getCellClass: function () {
+    return mazeUtils.isBeeSkin(this.props.skin) ? BeeCell : Cell;
+  },
+
+  changeSelection: function (row, col) {
+    this.setState({
+      selectedRow: row,
+      selectedCol: col
+    });
+  },
+
+  onCellChange: function (newSerializedCell) {
+    var row = this.state.selectedRow;
+    var col = this.state.selectedCol;
+
+    // this is technically a violation of React's "thou shalt not modify
+    // state" commandment.
+    // TODO figure out a clean way to do this purely with setState
```

Comment:
The problem here being that we're change one of the items in cells before we even call setState? I don't see any great solutions to that short of starting to use a library like immutablejs, or cloning our entire array of cells, modifying the single cell in the clone, and calling setState.

Another solution might be to present our state as an object, i.e.
```
state = {
  cell_0_0: cells[0][0],
  cell_0_1: cells[0][1]
}
```
at which point we can do `this.setState({cell_0_0: newCell})`
You would also probably want to provide a helper getCell(row, col) => { this.state['cell_' + row + '_' + col] }`
However, that mostly feels unnatural, and I don't know that it's worth it.

I think I'm okay with leaving this as is, as long as we leave in a comment explaining what's happening.